File: utils/src/main/java/jetbrains/exodus/util/MathUtil.java
/*
 * Copyright ${inceptionYear} - ${year} ${owner}
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.util;

public class MathUtil {

    /**
     * @param i integer
     * @return discrete logarithm of specified integer base 2
     */
    public static int integerLogarithm(final int i) {
        return i <= 0 ? 0 : Integer.SIZE - Integer.numberOfLeadingZeros(i - 1);
    }

    /**
     * @param l long
     * @return discrete logarithm of specified integer base 2
     */
    public static long longLogarithm(final long l) {
        return l <= 0 ? 0 : Long.SIZE - Long.numberOfLeadingZeros(l - 1);
    }

    /**
     * Returns the sum of {@code a} and {@code b} unless it would overflow or underflow in which case
     * {@code Long.MAX_VALUE} or {@code Long.MIN_VALUE} is returned, respectively.
     */
    public static long saturatedAdd(long a, long b) {
        long naiveSum = a + b;
        if (((a ^ b) < 0) | ((a ^ naiveSum) >= 0)) {
            // If a and b have different signs or a has the same sign as the result then there was no
            // overflow, return.
            return naiveSum;
        }
        // we did over/under flow, if the sign is negative we should return MAX otherwise MIN
        return Long.MAX_VALUE + ((naiveSum >>> (Long.SIZE - 1)) ^ 1);
    }
}

File: utils/src/main/java/jetbrains/exodus/util/MathUtil.kt
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.util

object MathUtil {
    /**
     * @param i integer
     * @return discrete logarithm of specified integer base 2
     */
    fun integerLogarithm(i: Int): Int {
        return if (i <= 0) 0 else Integer.SIZE - Integer.numberOfLeadingZeros(i - 1)
    }

    /**
     * @param l long
     * @return discrete logarithm of specified integer base 2
     */
    @JvmStatic
    fun longLogarithm(l: Long): Long {
        return if (l <= 0) 0 else (java.lang.Long.SIZE - java.lang.Long.numberOfLeadingZeros(l - 1)).toLong()
    }
}
